working/extension_structs/overview.md

```diff
@@ -0,0 +1,815 @@
+support wrapper-less views on an object.  This section describes how extension
+structs look and behave, largely by example.  In many cases, alternatives or
+extensions to the core proposal are described in line.
+
+## Extension structs by example
+
+Extension structs are targeted at relatively niche uses where you wish to define
+a set of statically dispatched methods layered on top of an underlying
+representation, without introducing a wrapper object.  A canonical use case
+driving this design is to be able define a Dart typed interface for methods on a
+Javascript object, providing a wrapperless interoperation capability.
+
+
+```dart
+extension struct Window(JSObject o) {
+  // Signatures provided here for methods to be delegated to the underlying
+  // JSObject
+  external bool get closed;
+  // etc
+}
+
+void test(Window w) {
+  // Window methods can be called using Dart syntax
+  if (w.closed) {...}
+
+  // Windows are represented as the underlying object
+  assert(w is JSObject);
+
+  // Windows can be cast to the underlying object
+  JSObject o = w as JSObject;
+
+  // Extension struct types are reified as the underlying representation type.
+  List<Window> l = [w];
+  assert(l is List<JSObject>);
+}
+```
+
+Extension structs are also useful for providing a lightweight facade over an
+existing type.
+
+```dart
+// Natural numbers
+extension struct Nat(int _x) {
+  Nat(int x) : assert(x >= 0), _x = x;
+  Nat.zero() : _x = 0;
+
+  Nat get succ => Nat(_x+1);
+  Nat plus(Nat other) => Nat._x + other._x;
+
+  // Override the underlying isNegative operation (incorrectly)
+  bool get isNegative => true;
+}
+
+void test() {
+  var n1 = Nat(3);
+  var n2 = Nat.zero();
+  assert(n2.succ.succ.succ == n2);
+
+  //The underlying representation is still as an int
+  assert(n1 is int);
+
+  // The static type is used to dispatch the method calls
+  assert(n1.isNegative);
+
+  // If the static type is lost, dispatch goes to the underlying object.
+  dynamic d = n1;
+  assert(!d.isNegative);
+}
+```
+
+Extension structs may delegate members to the underlying field.  They may also
+implement interfaces, but only if the underlying field implements the interface.
+
+```dart
+// Natural numbers
+extension struct Nat(int _x) implements Comparable<num> {
+  // Constructors etc as above
+
+
+  bool get isEven;  // Abstract definition delegates to _x.isEven
+}
+
+void test() {
+  var n = Nat(2);
+  // Same as 2.isEven
+  assert(n.isEven);
+  // Comparable interface allows access to the Comparable methods on int
+  assert(n.compareTo(2) == 0);
+
+  // Since Nat implements Comparable<num>, it may be assigned to it
+  Comparable<num> c = n;
+  // The representation is still as an integer
+  assert(c is int);
+}
+```
+
+Extension structs have no inheritance.
+
+```dart
+// Static error
+extension struct PositiveNumber(int _x) extends Nat {...}
+```
+
+Extension structs do not define a signature, and hence cannot be implemented.
```

Comment:
:+1: We definitely want to disallow implementing an interface from a extension struct.

That said, for JSInterop, we are also interested in some form of conformance checking. This is especially useful with mocking. I'll share a few examples here, but I am not sure this is a problem we can solve at this time.

Consider this code:
```dart
extension struct Foo(JSObject o) {
  external bool m1(int x);
}
```
Or the equivalent expansion:
```dart
extension struct Foo(JSObject o) {
  bool m1(int x) => js_util.callMethod(o, 'm1', [x]);
}
```

We'd like to create mocks in Dart that get exported at the JS level, so they can be a valid JSObject to that extension struct (in this case, `Foo` expects the underlying `JSObject` to have certain JS properties and methods).

Our current proposal is to say:
```dart
class MyFooMock {
  bool m1(int y) => false;
}
...
Foo mock = mockify<MyFooMock, Foo>(MyFooMock());
```

The type-parameters to mockify give us the magic we need to take `MyFooMock` and shape it as a `JSObject` with the expectations of `Foo`. 

We'd like that connection to be more explicit in the syntax, and eventually have a static error if `MyFooMock` doesn't conform to the extensions declared in `Foo`. We don't want `MyFooMock implements Foo`, but if If `Foo` had an implied interface `ExtensionStructInterface<Foo>`, we would like to say `MyFooMock implements ExtensionStructureInterface<Foo>`.

What makes this harder, is that it's not true that Foo and the mock need to match, but how Foo uses the JSObject and the mock do. So to make this example more complicated:
```dart
extension struct Foo(JSObject o) {
  @JS('m2')
  external bool m1(int x);

  int m3() => 3;
}
```
Or the equivalent expansion:
```dart
extension struct Foo(JSObject o) {
  bool m1(int x) => js_util.callMethod(o, 'm2', [x]);
  int m3() => 3;
}
```

The ideal mock class should be:
```dart
class MyFooMock {
  bool m2(int x) => false;
}
```

:exploding_head: 

/cc @srujzs

Reply:
Yeah, I don't really see a nice way to get this.  One thing you could do with my proposal (almost) is define `FooInterface` as an abstract class, and say that `Foo implements FooInterface`, and then you can also have `class MyFooMock implements FooInterface` .  The `m1` -> `m2` mapping you'd be on your own for... :) but maybe you could mark that in the interface.  To get this from my proposal though, you need to make two adjustments (I think)
  - You need to allow members from interfaces to be "overridden" in the extension struct (there's nothing problematic with this semantically, but as I discuss in the proposal it has slightly surprising semantics)
  - You need to say that JSObject implements all interfaces (or at least interfaces marked in certain ways), since I require that the "on type" implements the interface as well.

Reply:
> What makes this harder, is that it's not true that Foo and the mock need to match, but how Foo uses the JSObject and the mock do. (etc.)

I mentioned one complication for this earlier on how JS members may have characters in their name that aren't supported in Dart, but maybe it doesn't happen often. Either way, you'd still need a `mockify` call to wrap it and all that, so it's just slightly cleaner. One reason I wanted to match the Dart names is that the interfacing we describe is actually possible (since `MyFooMock.m1` implements `Foo.m1`, whereas `MyFooMock.m2` does not implement `Foo.m1`, since the language does not care about our renaming annotations).

> One thing you could do with my proposal (almost) is define FooInterface as an abstract class...

Yeah, you'd still need to ensure that `MyFooMock` implements all the members in `Foo`, unless you start adding weird restrictions that `MyFooMock` doesn't add any new members.

P.S. I may be reading the proposal wrong, but wouldn't we need `JsObject` to implement the interfaces in order for an extension struct to implement those interfaces?

It's not *that bad* if we can't get interfaces for extension structs, it's just a *slightly* worse user experience for writing mocks.

Reply:
> P.S. I may be reading the proposal wrong, but wouldn't we need JsObject to implement the interfaces in order for an extension struct to implement those interfaces?

@srujzs I'm not entirely sure I understand your question - are you asking about mocking, or in general?  In general, yes, this is a question that I wanted to get input from you all.  It seems "reasonable" to me, to say that JSObject implements all interfaces for the purposes of this conformance check.  So then you can write something like `view class MyJSArray(JSObject o) implements List<dynamic>;`.  It does of course allow you to do all sorts of nasty things: you can pass in any old JSObject to create a MyJSArray, and if you assign it to a `List<dynamic>` chaos ensues.  You can add validation in the constructor if you want (in fact, we could probably even specify this in general), but casts can still get around this.

Reply:
Not sure how I missed your second bullet point earlier - you already answered it there, sorry. Having `JsObject` implement *all* interfaces is indeed chaos, and I think that would be worse than supporting the interfacing behavior.

It seems like it's fundamentally hard to get this behavior (where an extension struct can provide an interface that is used for another class/struct). Picking your brain here a bit I guess, but why do we want to disallow structs defining a signature?

Reply:
I don't really know what it means for an extension struct to define a proper interface.  How does a class implement that interface?  I can in principle allow extension structs to implement other extension structs, but I'm not sure that really helps.

We should talk about how you want to deal with JsObject.  It feels to me like we should be able to find a way to mark specific uses of JsObject as implementing specific interfaces, if we want to do that.  I can imagine something like:

Reply:
I'm guessing there was supposed to be more to this comment from the last sentence.

> I don't really know what it means for an extension struct to define a proper interface. How does a class implement that interface?

I'd have to imagine it's some synthetic class that's created for that view that can be implemented (like Siggi mentions), but that seems heavy and ugly. Since classes can't implement structs, boxing (with your proposed `.struct`) doesn't help here either.

Going back a bit here to Siggi's comment:

> We don't want `MyFooMock implements Foo`

*Why* don't we want this, again? I'd expect the class to conform to every member signature (and if needed, the singular field's signature as well), just like with classes, but I feel like I'm missing something.

> I can in principle allow extension structs to implement other extension structs, but I'm not sure that really helps.

Yeah, the lack of state from extension structs (since we don't have fields) would make mocks less useful anyways, so I don't think this is needed.

Reply:
> I'd have to imagine it's some synthetic class that's created for that view that can be implemented (like Siggi mentions), but that seems heavy and ugly.

Sorry, I still don't understand how this works.  views use static dispatch.  I can perfectly well allow you to say that `Foo` implements `MyStruct`, but when I assign a `Foo` to a `MyStruct`, the methods that get called will not be the `Foo` methods, but the `MyStruct` methods.  That's what static dispatch means.

> I'm guessing there was supposed to be more to this comment from the last sentence.

Sorry, I meant to delete that last bit.  But to close it, we could do something like:

```
// Magic marker typedef
typedef JSObjectImplements<T> = T;
struct Window(JSObject o) implements JSObjectImplements<List<int>> {}
```

or really any number of other things.  You could go the other way around and do:

```
typedef JSObjectImplements<T> = JSObject
struct Window(JSObjectImplements<List<int>> o) implements List<int> {}
```
though this one doesn't generalize well to the n-ary case.

There are other things we could brainstorm.

Reply:
>> We don't want MyFooMock implements Foo

> _Why_ don't we want this, again? I'd expect the class to conform to every member signature (and if needed, the singular field's signature as well), just like with classes, but I feel like I'm missing something.

My point was mainly to make a distinction that static struct/views do not have a virtual interface. I fear that writing `MyFooMock implements Foo` can lead to confusion because developers will expect an assigment like `Foo x = MyFooMock()` to work. I'd be more comfortable if Foo's implied interface is distinguished from the type name syntactically somehow. For example, I'd be OK with `MyFooMock implement Foo.interface` or `MyFooMock implements ImpliedInterface<Foo>` (that's what I meant by `ExtensionStructureInterface<Foo>` in the earlier example)

Reply:
> I can perfectly well allow you to say that Foo implements MyStruct, but when I assign a Foo to a MyStruct, the methods that get called will not be the Foo methods, but the MyStruct methods.

Right, I guess the problem here is assignability as Siggi mentions. I guess `implements` is the wrong way to think about this, as we only what the signature conformance side of that feature, not assignability or subtyping.

And okay, I agree that explicitly mentioning `.interface` or something similar would avoid the confusion where users think they could subtype.

I do not see any plausible usage for this besides specifically for interop mocking (which we can check anyways, but just at a later stage unfortunately, unless we can maybe leverage the analyzer), so I'm inclined to agree that this is just a small very specific nice-to-have that I'm okay with dropping. After all, why would anyone want to make sure the signatures match if they never plan on using one type for the other? :)

Reply:
> unless we can maybe leverage the analyzer

Possibly a candidate for a lint.